### PR TITLE
Fix for Issue #191

### PR DESCRIPTION
**Pull request by codeinteger6**
Here is the fix for Issue #191. The problem was caused due to Discord's character limit, if the user has too many roles the thread creation process won't work properly. This PR sets a limit on how many roles the bot can display before it says "Too many roles to show" therefore not exceeding Discord's character limit. I have set the value to 21, meaning it can display up to 20 roles before it says too many roles.

Initially I tried setting the value to 41, but then the embed message of where the user info and roles are shown wasn't sending at all, so I reduced the value to 21 and now it's working fine.

### CODE REVIEW

**Comment by codeinteger6** [comment]
It is similar to what Dyno does when using whois on a user with a lot of roles.

**Comment by fourjr** [comment]
i dont think that is relevant to modmail?

**Comment by codeinteger6** [comment]
It will at least solve the issue, so best to deploy this now, if needed changes can be made later

**Comment by AmitGamingKing** [comment]
For your question @fourjr if any user has more roles, then just change value in `if` condition as of now it is set to 41 roles including the `@everyone` role you can change it to a specific limit but as we tried it didn't work for more roles and same problem was happening so this is a basically a quick fix! However devs will surely work on this thing in future commits...

**Comment by fourjr** [comment]
I suggest checking the size of the description and ensure it does not cross the limit.

**Comment by kyb3r** [comment]
Instead of using a role number limit it would be more robust to use a character limit.

**Comment by codeinteger6** [comment]
Will see what can be done and report back.

**Comment by codeinteger6** [comment]
After doing some extended testing and modifications, here is the final version of the thread.py file.

Some things to point out:

1. In non-separate server mode, the amount of characters in the Discord role name does not matter as it uses IDs to take up the character limit. I tried setting the value to 46, but it was exceeding the 1024 character limit. 41 is fine though (meaning 40 roles will be displayed excluding everyone role).

2. In separate server mode, this is where character count is important unlike the non separate mode where it uses mentions so only IDs take up the character count. The bot can handle up to 1024 characters in a single embed. But at the same time, need to make space for other characters in the embed so I decided to allow up to 800 characters (including roles, commas and spaces in total) that can be displayed, so there is enough room for the other characters in the embed, including the full username which can go up to 32 characters.

3. A role count is also added.

This is the best that could be done, hope this is sufficient.

**Comment by fourjr** [comment]
Wouldn't a simple `value[:1000] + '...'` work?

**Comment by codeinteger6** [comment]
Non separate server mode the character count does not matter. Because it counts the Discord IDs only in the character count not the role name as it uses mentions. It can display up to 40 without problems, 45 roles exceeds that limit.

For separate server mode it's set to up to 800 characters for the role list, it needs space for other things like name, join date, number of threads, ID and time in the embed as well.

But I believe that the thread.py file currently is ready to be merged. If necessary, changes can be made later but it will at least not cause problems for users with so many roles in certain servers. I tried to set the limits as high as it can while taking everything into account.

**Comment by codeinteger6** [comment]
I need to check for separate server mode again, don't merge now. I will come back with an updated file if needed.

**Comment by codeinteger6** [comment]
You were right. I just checked the Discord developer docs and the field.value, which is where all the roles are displayed can't exceed more than 1024 characters. I edited the file for both separate and non-separare server modes and now I'm fully confident it is ready to merge and deploy the update.

File: core/thread.py [hunk 1]
```diff
@@ -645,15 +645,30 @@ def _format_info_embed(self, user, log_url, log_count, color):
...
+                if len(charCounter) <= 1024:
+                    role_names = ', '.join(r.name for r in roles
+                                            if r.name != "@everyone")
+                else:
+                    role_names = "Due to Discord limitations the bot can't display roles that contain more than 1024 characters in total."
             else:
-                role_names = ' '.join(r.mention for r in roles
-                                      if r.name != "@everyone")
+                for r in roles:
+                    count = count + 1
```
Comment (fourjr) [inline core/thread.py]: count += 1

File: core/thread.py [hunk 2]
```diff
@@ -680,7 +695,8 @@ def _format_info_embed(self, user, log_url, log_count, color):
                                 value=member.nick,
                                 inline=True)
             if role_names:
-                embed.add_field(name='Roles',
+                count = count - 1
```
Comment (fourjr) [inline core/thread.py]: count -= 1

File: core/thread.py [hunk 3]
```diff
@@ -645,15 +645,30 @@ def _format_info_embed(self, user, log_url, log_count, color):
...
+        count = 0
         if member:
             separate_server = self.bot.guild != self.bot.modmail_guild
             roles = sorted(member.roles, key=lambda c: c.position)
             if separate_server:
-                role_names = ', '.join(r.name for r in roles
-                                       if r.name != "@everyone")
+                for r in roles:
+                    count = count + 1
+                charCounter = ', '.join(r.name for r in roles
```
Comment (fourjr) [inline core/thread.py]: use snake_case

File: core/thread.py [hunk 4]
```diff
@@ -645,15 +645,30 @@ def _format_info_embed(self, user, log_url, log_count, color):
...
+                                            if r.name != "@everyone")
+                if len(charCounter) <= 1024:
+                    role_names = ', '.join(r.name for r in roles
+                                            if r.name != "@everyone")
+                else:
+                    role_names = "Due to Discord limitations the bot can't display roles that contain more than 1024 characters in total."
```
Comment (fourjr) [inline core/thread.py]: i think it's best to show what we can & put ellipsis after

**Comment by kyb3r** [comment]
> Wouldn't a simple `value[:1000] + ('' if len(value) <= 1000 else '...')` work?

This would mess up role mentions, I think it'd be better to iteratively concatenate role names/mentions until the length of the string is just under 1024.

e.g. 
```py
roles = []

for i, r in enumerate(user.roles):
    fmt = r.name if sep_server else r.mention
    if sum(len(r) for r in roles) + len(fmt) > 1000:
        roles.append('...')
        break
    roles.append(fmt)

seperator = ', ' if sep_server else ' '
roles = seperator.join(roles)
```

**Comment by fourjr** [comment]
What about commands.Paginator? It implements that iirc

**Comment by AmitGamingKing** [comment]
@fourjr My opinion is you create a new pull request by your own implementation and show that to us would be great for you & everyone :D

**Comment by codeinteger6** [comment]
What else can be done differently? My [latest commit](https://github.com/kyb3r/modmail/pull/200/commits/9fb31e39b220788904dcec949c9c05dc28f33bd6) I have tested it in both separate and non-separate mode and I believe it is fine the way it is. The problem was the field.value where the role lists are displayed can only go up to 1024 characters according to Discord developer docs. So i made it only display up to 1024 characters if it goes more than that it will say can't show roles but it will at least not crash and cause problems for some servers.

Test it and see if anything needs to be changed but I'm pretty sure there's nothing else that need to be done and ready for deployment.

**Comment by codeinteger6** [comment]
And before starting to edit the thread.py file I checked what the original problem was by reproducing the bug and seeing what comes up in the logs:

```
2019-03-19T08:39:16.782389+00:00 app[worker.1]: future: <Task finished coro=<Thread.setup() done, defined at /app/core/thread.py:79> exception=HTTPException('BAD REQUEST (status code: 400): Invalid Form Body\nIn embed.fields.0.value: Must be 1024 or fewer in length.')>
2019-03-19T08:39:16.782418+00:00 app[worker.1]: In embed.fields.0.value: Must be 1024 or fewer in length.
```
You can see it was the field.value exceeding 1024 characters.

**Comment by fourjr** [comment]
We don't develop modmail just for "it to work" and getting the fastest solution. It is also for style and the cleanest possible code. If you feel that you want to stick to your messy code, feel free to do so in your own fork.

**Comment by codeinteger6** [comment]
Here's a cleaner code

**Comment by kyb3r** [comment]
A bit late but I implemented this in 2db8b854caa14aa2f7434a28fa636a865445c0e3

**Pull request closed without merging**